docket/config.py

```diff
@@ -0,0 +1,153 @@
+from __future__ import print_function
+# A central location for the config dictionary
+#   see application for initialization
+import yaml
+import os
+import logging
+
+levels = {
+    'debug'  :logging.DEBUG,
+    'info'   :logging.INFO,
+    'warn'   :logging.WARNING,
+    'warning':logging.WARNING,
+    'error'  :logging.ERROR,
+    'crit'   :logging.CRITICAL,
+}
+log_level = lambda x: levels.get(x, logging.INFO) if type(x) is str else int(x)
+
+class Config:
+    """ finds, loads, and provides access to config from FILENAME.yaml
+        Verifies configuration default settings are uniform
+        provides a uniform logging format
+    """
+    loggers = {}
+    logger  = None
+    formatter = None
+    log_level = None
+    file_handler = None
+    config = {
+        'WEB_ROOT': '/',
+        'SPOOL_DIR': '/var/spool/docket',   # overridden by environment
+        'SECRET_KEY': 'CHANGE_THIS',        # overridden by environment
+        'CELERY_TASK_IGNORE_RESULT': True,
+        'LOG_MSG_FORMAT': '[%(processName)-8s] %(message)s',
+        'LOG_DATE_FORMAT': '%Y-%m-%dT%H:%M:%SZ',
+        'LOG_DATE_UTC': True,
+        'LOG_LEVEL': 'info',
+        'LOG_FILE_BACKUPS' : 7,
+        'LOG_FILE_MSG_FORMAT': '%(asctime)s[%(processName)-8s] %(message)s',
+        'LOG_FILE_DATE_FORMAT': '%Y%jT%H:%M:%S',
+        'LOG_FILE_LEVEL': 'info',
+    }
+    _default = config.copy()
+
+    @classmethod
+    def load(cls, filename=None, flask_app=None, logger=None):
+        """ Set Config from a yaml file:
+            filename a config file: search for the file in several directories
+        """
+
+        import itertools
+        app_root = flask_app.root_path if flask_app else '/opt/docket/docket'
+        fs_root  = os.path.abspath(os.sep)
+        paths = itertools.product(
+            ( filename,                     # filename.yaml
+             'docket',                      # docket.yaml
+            ),
+            [ tuple(),                      # <empty> - filename can be any valid path
+             (fs_root, 'etc', 'docket'),    # /etc/docket/filename
+             (fs_root, 'etc',),             # /etc/filename
+#            (os.curdir, os.pardir, 'conf'),# CWD/../conf/filename
+             (app_root, 'conf'),            # APP/conf/filename
+             ('conf',)]                     # ./conf/filename
+        )
+        for f,p in paths:
+            if not f:
+                continue
+            filepath = os.path.join(*(p + (f,)) )
+            filepath = os.path.expanduser(filepath)
+            if os.path.isfile(filepath):
+                break
+            elif os.path.isfile(filepath+'.yaml'):
+                filepath += '.yaml'
+                break
+            print("Config: >{}< not found".format(filepath))
+
+        # load the config file and update our class config data
+        with open(filepath) as f:
```

Comment:
What happens if the config file doesn't exist at all?

Reply:
An exception is logged after a list of 'Config >PATH< not found' messages.

Considering that we currently expect stenographer instances to be found in that config file,
I hope that missing it is a reasonably understandable error for prospective administrators.